torch/csrc/jit/autodiff.h

```diff
@@ -7,10 +7,80 @@
 
 namespace torch { namespace jit {
 
-// This function mutates the given graph (which should only contain a single stage)
-// by appending nodes of a next stage, which computes the Jacobian-vector product
-// (aka backward) of inputs to the first stage w.r.t. the outputs of the first stage.
-void differentiate(std::shared_ptr<Graph>& graph);
+struct Capture {
+  enum class Kind {Input, Output};
+  Capture(Kind kind, std::size_t offset)
+    : kind(kind), offset(offset) {}
+  Kind kind;
+  std::size_t offset;
+};
+
+using value_list = std::vector<Value*>;
+// Example showcasing how Gradient is constructed:
+//
+// Let's assume we have a function f, `m` and `n` do not require grad
+// (`n` can depend only on `m`):
+//   y, n = f(x, m)
+//
+// Now, let's assume that the reverse of f (called f') needs to use values of `x`, `t` and `y`.
+// `t` is an intermediate value produced in the body of f, and let's assume that it requires
+// grad too.
+//
+// In this case differentiate(f) will return this:
+//   y, n, t = f(x, m)        // `t` is appended to the output list
+//   dx = f'(x, t, y, dy, dt) // No `dm` or `dn` because they do not require gradient
+//                            // All needed values from f are prepended to the input list
+//
+//   f_real_outputs = 2       // Only first two outputs were present in f originally
+//   df_input_captures = {O0, O2, I0} // Order matches the prefix of inputs to df
+//                        y   t   x
+//   df_input_vjps = {0, 2}   // i.e. connect grad_fn of y and t variables produced by f,
+//                    y  t    // with y's output_nr = 0 and t's output_nr = 1
+//   df_output_vjps = {0}     // i.e. connect next_function[0] of grad_fn to x's (grad_fn, output_nr).
+struct Gradient {
+  std::shared_ptr<Graph> f;
+  std::shared_ptr<Graph> df;
+
+  // Describes how to construct outputs of f from what its graph will return.
+  // This is necessary because some trailing outputs are intermediates produced
+  // only to be saved for df (and should be ignored).
+  std::size_t f_real_outputs;
+
+  // df inputs are split into two sections: captures are vjps (aka grad_outputs).
+  // Captures are values the need to be saved when f is run. We handle inputs
+  // specially, because this allows us to avoid adding extra vjps as df inputs.
+  // VJPs are "seeds" for the gradient computation given for each input capture
+  // of an Output kind.
+  std::vector<Capture> df_input_captures;
+  std::vector<std::size_t> df_input_vjps; // Offsets into f's outputs.
```

Comment:
I'm not a fan of the member name `df_input_vjps` because no where does it cue that this is actually a vector of *indexes* into the outputs array (similar with `df_output_vjps`).

Reply:
I'm not a huge fan, but it's ok. It tells you vjps for which outputs are inputs to df. I can't find a better name